src/Layers/xrRender/BufferUtils.h

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+u32 GetFVFVertexSize(u32 FVF);
+u32 GetDeclVertexSize(const VertexElement* decl, DWORD Stream);
+u32 GetDeclLength(const VertexElement* decl);
+
+struct SDeclaration;
+
+void ConvertVertexDeclaration(u32 FVF, SDeclaration* decl);
+void ConvertVertexDeclaration(const VertexElement* dxdecl, SDeclaration* decl);
+void ConvertVertexDeclaration(const xr_vector<VertexElement>& declIn, xr_vector<InputElementDesc>& declOut);
+
+namespace BufferUtils
+{
+HRESULT CreateVertexBuffer(VertexBufferHandle* pBuffer, const void* pData, UINT DataSize, bool bImmutable = true);
+HRESULT CreateIndexBuffer(IndexBufferHandle* pBuffer, const void* pData, UINT DataSize, bool bImmutable = true);
+HRESULT CreateConstantBuffer(ConstantBufferHandle* ppBuffer, UINT DataSize);
+};
+
+class VertexStagingBuffer
+{
+public:
+    VertexStagingBuffer();
+    ~VertexStagingBuffer();
+
+    void Create(size_t size);
+    void Destroy();
+
+    bool IsValid() const;
+    void* GetHostPointer();
```

Comment:
It is better to rename this function to `Lock`.
```suggestion
    void* Lock();
```

Reply:
I agree

Reply:
Changed to `Map`/`Unmap` pair to be consistent with forthcoming streaming buffers

Reply:
It's even more complicated now and should be made easier.

Reply:
It is absolutely obvious to call `Unlock()` if you called `Lock()`.
It is less, but obvious to call `Unmap()` if you called `Map()`.

But it is non-obvious to call `Flush()` if you called `Map()`

Reply:
This complexity was added with purpose. In the case of skinned meshes loading, we will need access to buffer data again.
I can add all the code reworked but this will increase the PR. Or can share my local branch with all geometry stuff reworked

Reply:
Agree with Lock/Unlock inconsistency

Reply:
> This complexity was added with purpose. In the case of skinned meshes loading, we will need access to buffer data again.
I can add all the code reworked but this will increase the PR. Or can share my local branch with all geometry stuff reworked

Then we should remove `Unmap()` function and rename `Flush()` to `FlushAndUnmap()`. This is very consistent:
1. Linked with `Map()` functions (function name tells you that you should call it)
2. Corresponds with what the functions does on DX10+/OpenGL. (It flushes the buffer and then discards the host buffer)